WIP: Staged layer creation

Luap99 opened the pull request:
<!--- Please read the [contributing guidelines](https://github.com/containers/container-libs/blob/main/CONTRIBUTING.md) before proceeding --->

Luap99 commented:
Podman PR https://github.com/containers/podman/pull/27251 and the buildah test PR https://github.com/containers/buildah/pull/6414 from the bot both look good so that means we can remove the special case from ApplyDiff() in overlay I think, ref https://github.com/containers/podman/issues/25862#issuecomment-3361659429

I still need to work on the actual feature here though to extract while the store in unlocked.

mtrmac commented on storage/store.go:
```diff
@@ -3177,7 +3177,7 @@ func (s *store) ApplyStagedLayer(args ApplyStagedLayerOptions) (*Layer, error) {
 		// assume layer creation and applying a staged layer are separate steps. Production pull code always uses the
 		// other path, where layer creation is atomic.
-		return layer, rlstore.applyDiffFromStagingDirectory(args.ID, args.DiffOutput, args.DiffOptions)
+		return layer, rlstore.applyDiffFromStagingDirectory(layer, args.DiffOutput, args.DiffOptions)
```
Very non-blocking

I’m a _bit_ worried about trusting the out-of-`layers.go` code to keep a correct locked value of `layer` here; ISTR we mostly keep “this data is kept locked and consistent” to the `layers.go` layer of the call stack; in `store.go`, a `Layer` is more of a “possibly-already-obsolete snapshot of data”.

Technically this clearly works (we are holding the RW layer store lock here).

As a general long-term direction, I’d like to get rid of the “if !layerExists { report error }; doOperation” pattern, in favor of a direct ”doOperation; handle layerDoesNotExistError“. This applies to the calls from `store.go` to `layers.go`, mostly just because the latter is shorter; and it applies much more to the calls from external callers to `store.go`, because in that case two calls means two expensive locking operations, and possibly the situation changing in the meantime. (Also, the `Get` calls also serve as a “resolve name or ID prefix to a full ID” step — _ideally_ that should not be happening at all, after we process/resolve CLI inputs. But changing that at the external c/storage API surface is somewhat risky.)

So, I’d prefer `rwLayerStore.applyDFSD` to continue to accept an ID, and maybe `layers.go` can add an intermediate function with a `*Layer`.

But then again, see the comment just above — this is ~mostly dead code. That… means _both_ that it is not worth investing much into, and that it is at higher risk of accidentally being broken.

Also, there’s `.create` already using a pointer, so this is by no means a hard rule.

mtrmac reviewed (commented):
ACK, simplifying `ApplyDiff` this way does look correct. (I didn’t carefully look at the `tempdir` addition yet.)

Luap99 commented on storage/layers.go:
```diff
@@ -1474,6 +1479,41 @@ func (r *layerStore) create(id string, parentLayer *Layer, names []string, mount
+	applyDiffTemporaryDriver, ok := r.driver.(drivers.ApplyDiffStaging)
+	if ok && diff != nil {
+		// CRITICAL, this releases the lock so we can extract this unlocked
+		r.stopWriting()
```
I guess I might have been to focused to make proper quick ID and name conflict lookup first before doing the expensive lookup to "fail fast" when possible.
I guess design wise it makes sense to push this all the way up the stack. I do agree that unlock/lock patter is quite dangerous and I have seen it fail to many times in podman already so if we can avoid it then we should do that

Luap99 commented:
@mtrmac FYI I have not really addressed most of your comments yet, I am just trying to push things to see how much things break. Still seeing plenty of test failures.

Issue 1 I see is that I just use the 700 permission from the tmpdir due the rename instead of the proper diff dir creation permissions that are in the driver.create() code
```
	diff := path.Join(dir, "diff")
	if err := idtools.MkdirAs(diff, forcedSt.Mode, forcedSt.IDs.UID, forcedSt.IDs.GID); err != nil {
		return err
	}

	if d.options.forceMask != nil {
		st.Mode |= os.ModeDir
		if err := idtools.SetContainersOverrideXattr(diff, st); err != nil {
			return err
		}
	}
```

Not sure if I should expose that into the tmpdir creation logic, I guess that makes the most sense since only the dirver should now the exact permission that should be used?


---

Second problem I see are timeouts (in parallel running tests) which I guess mean I added a deadlock situation?
https://api.cirrus-ci.com/v1/artifact/task/5906611702595584/html/sys-podman-debian-13-rootless-host-sqlite.log.html

I guess looking at the code this unlock/lock again thing I did is indeed completely broken and unsafe due ABBA deadlock, i.e. in putlayer we also hold the containerStore lock so only unlocking the layer store makes it possible that another process can get the layer lock and then blocks on the still gold container store thus both process handing forever.
I haven't checked all the code paths but I guess with the locking order requirements what I did is basically impossible to achieve anyway and I have to indeed move this out to before we get the lock?

mtrmac commented:
> Issue 1 I see is that I just use the 700 permission from the tmpdir due the rename instead of the proper diff dir creation permissions that are in the driver.create() code

> Not sure if I should expose that into the tmpdir creation logic, I guess that makes the most sense since only the dirver should now the exact permission that should be used?

I think that could work.

I was thinking `StageAddition` does not actually need to _create_ (`os.Create`/`os.Mkdir`) the `tmpAddPath` at all. All of that happens inside a lock-protected `td.tempDirPath`, so there is ~nothing special, that I can see, about populating `tmpAddPath` — the provided `callback` can create the staged item without any help. (That could also mean `StageDirectoryAddition` and `StageFileAddition` could be consolidated into one. And I’m not immediately sure we _need_ a `callback` — `StageAddition` could return a `newStagingPathToPopulate` — but I also didn’t now carefully re-read the `tempdir` package.)

---

> I haven't checked all the code paths but I guess with the locking order requirements what I did is basically impossible to achieve anyway and I have to indeed move this out to before we get the lock?

Per the locking hierarchy documented at the top of `store`, I think you’re right here.

giuseppe commented on storage/layers.go:
```diff
@@ -2417,6 +2448,78 @@ func createTarSplitFile(r *layerStore, layerID string) (*os.File, error) {
+			Mappings: idtools.NewIDMappingsFromMaps(layerOptions.UIDMap, layerOptions.GIDMap),
+			// FIXME: What to do here? We have no lock and assigned label yet.
+			// Overlayfs should not need it anyway so this seems fine for now.
+			MountLabel: "",
```
I think this is correct

giuseppe commented on storage/drivers/overlay/overlay.go:
```diff
@@ -2369,21 +2369,91 @@ func (d *Driver) DifferTarget(id string) (string, error) {
+
+	sa, err := t.StageAddition()
+	if err != nil {
+		return nil, nil, -1, err
```
missing t.Cleanup?

mtrmac commented on storage/drivers/driver.go:
```diff
@@ -299,6 +299,15 @@ type DriverWithDiffer interface {
+// ApplyDiffStaging is an interface for driver who can apply the diff without holding the main storage lock.
+// This API is experimental and can be changed without bumping the major version number.
+type ApplyDiffStaging interface {
+	// StartStagingDiffToApply applies the layer in a temporary directory. This can be done without holding the storage lock.
```
Eventually please copy & paste the detailed description here, e.g. about the unusual `CleanupTempDirFunc` return value.

mtrmac commented on storage/layers.go:
```diff
@@ -2398,37 +2441,106 @@ func (r *layerStore) ApplyDiff(to string, diff io.Reader) (size int64, err error
-func (r *layerStore) applyDiffWithOptions(to string, layerOptions *LayerOptions, diff io.Reader) (size int64, err error) {
-	if !r.lockfile.IsReadWrite() {
-		return -1, fmt.Errorf("not allowed to modify layer contents at %q: %w", r.layerdir, ErrStoreIsReadOnly)
+func createTarSplitFile(r *layerStore, layerID string) (*os.File, error) {
```
(Non-blocking: There’s also the `templateTSdata` case, where we could do a file-to-file copy without holding it in memory.)

mtrmac commented on storage/layers.go:
```diff
@@ -2481,62 +2594,105 @@ func (r *layerStore) applyDiffWithOptions(to string, layerOptions *LayerOptions,
+func applyDiffResultToLayer(r *layerStore, layer *Layer, layerOptions *LayerOptions, result *applyDiffResult) {
+	updateDigestMap(&r.bycompressedsum, layer.CompressedDigest, result.compressedDigest, layer.ID)
+	layer.CompressedDigest = result.compressedDigest
+	if layerOptions != nil && layerOptions.OriginalDigest != "" && layerOptions.OriginalSize != nil {
```
Isn’t this already handled in `applyDiff`?

mtrmac commented on storage/layers.go:
```diff
@@ -2398,37 +2441,106 @@ func (r *layerStore) ApplyDiff(to string, diff io.Reader) (size int64, err error
+	return sl.staging.CommitStagedLayer(layerID, sl.stagedLayer)
+}
+
+func applyDiff(layerOptions *LayerOptions, diff io.Reader, tarSplitFile *os.File, applyDriverFunc func(io.Reader) (int64, error)) (*applyDiffResult, error) {
```
(Explicitly warning that this might be called without any layer store locks held would be nice.)

mtrmac commented on storage/layers.go:
```diff
@@ -198,8 +199,37 @@ type DiffOptions struct {
+	// diff contains the tar archive, can be compressed
+	diff    io.Reader
+	staging drivers.ApplyDiffStaging
+	result  *applyDiffResult
```
Later: Please document what is set when / who is responsible for what, if it can’t be expressed by structure of the types.

mtrmac commented on storage/layers.go:
```diff
@@ -2398,37 +2441,106 @@ func (r *layerStore) ApplyDiff(to string, diff io.Reader) (size int64, err error
+			Mappings: idtools.NewIDMappingsFromMaps(layerOptions.UIDMap, layerOptions.GIDMap),
+			// FIXME: What to do here? We have no lock and assigned label yet.
+			// Overlayfs should not need it anyway so this seems fine for now.
+			MountLabel: "",
```
It’s _pretty_ likely I’m missing something (e.g. c/image always passes `""`), but AFAICT the value is always caller-supplied in `Store.PutLayer` and the like. So I don’t see that much of a difficulty in following the existing semantics.

If there _is_ a difficulty, please document in `StartStagingDiffToApply` that it must not be implemented by drivers that need `MountLabel` (or define a new `StartStagingDiffToApplyOpts` that make it impossible) and fail somewhere around here if that rule is violated.

Luap99 replied:
I guess you are right, I didn't really follow the purpose of this field.

In create there is 
```
if mountLabel != "" {
		selinux.ReserveLabel(mountLabel)
}
```

I don't know the side effects of this, is it fine to extract with a label that is not yet reserved? Looking closer into github.com/opencontainers/selinux suggest to me that the entire reserve logic might be broken already as it only depends on a in memory map which well just doesn't work at all with a the non daemon podman architecture. So I guess yeah I just pass the the caller label in here and don't worry about it.

mtrmac replied:
I missed the `ReserveLabel` call. The implementation of that _silently ignores conflicts_, so, I don’t think it can be really useful as is — if the label was not reserved previously, we could have a conflict and not notice it.

I don’t understand the whole thing… the ~only uses of label reservation for mount labels, to allocate a new value, seem to be:
- `drivers/vfs`, for some reason
- container creation (and, sure, that has various uses in Podman) — and that _does_ call `layerStore.Create` for the RW layer, but does not populate it with contents.
- `libpod.LabelVolumePath`, which allocates a label _and immediately releases it_ before using it (?!)

> it only depends on a in memory map which well just doesn't work at all with a the non daemon podman architecture.

That part _might_ (??) be fine, `layerStore.load` tries (again, ignoring errors) to reserve labels for pre-existing layers. Of course, that doesn’t account for any other callers of this API.

---

AFAICT the mount labels are _never_ set for _image_ layers in the Podman codebase, so all of this might be theoretical; they are set only for container layers.

It’s also very possible I’m missing something crucial about the design. @giuseppe?

---

Cleaning up / documenting the label reservation is, I think, out of scope — but I’d like to avoid making the hairball any larger. So, if it looks like we can’t safely handle non-empty label values, let’s add an extra condition, please.

mtrmac commented on storage/store.go:
```diff
@@ -1533,19 +1533,39 @@ func (s *store) putLayer(rlstore rwLayerStore, rlstores []roLayerStore, id, pare
+			}
+		}()
+		// FIXME: type case should be safe for now but really there should be a better way to do this
+		err = m.stageWithUnlockedStore(rlstore.(*layerStore), lOptions)
```
A dumb and inelegant way to do this would be to switch the parameters, `rlstore.stage…(m, …)`.

(I also wouldn’t mind too much getting rid of the interfaces entirely… OTOH I was searching for some way to make the layer store kind and lock holding state visible to the type system — I couldn’t _find_ a practical one with Go; still, getting rid of the interfaces would be a move in the opposite direction.)

mtrmac replied:
https://github.com/containers/container-libs/pull/378#issuecomment-3438325936 talks about ID mapping values from the parent layer — if I’m reading the code right, that does not yet happen.

mtrmac commented on storage/store.go:
```diff
@@ -1533,19 +1533,39 @@ func (s *store) putLayer(rlstore rwLayerStore, rlstores []roLayerStore, id, pare
+	var slo *stagedLayerOptions
+
+	if diff != nil {
+		m := newMaybeStagedLayerExtraction(diff, s.graphDriver)
```
Formally, this is racy WRT updates of `s.graphDriver` … while using `rlstore.driver` is not.

mtrmac commented on storage/drivers/overlay/overlay.go:
```diff
@@ -2369,31 +2369,91 @@ func (d *Driver) DifferTarget(id string) (string, error) {
+// CommitFunc is only set when there is no error returned and the int64 value returns the size of the layer.
+//
+// This API is experimental and can be changed without bumping the major version number.
+func (d *Driver) StartStagingDiffToApply(options graphdriver.ApplyDiffOpts) (tempdir.CleanupTempDirFunc, *tempdir.StageAddition, int64, error) {
```
A warning that this can run concurrently with any other operations on the driver would be nice (both here and in the interface definition).

mtrmac reviewed (commented):
Reading this commit by commit, this looks really great — the comments about documenting locking semantics etc. are basically the final polish.

(Note to self: Eventually it might be worth re-reading the final state as is, to check whether there is any opportunity to simplify.)

Around https://github.com/containers/container-libs/pull/378#discussion_r2427205226 and more recently with the parent’s mapping there was some tentative discussion about checking whether the layer exists before deciding to stage it — that’s still to be decided, I think. (In c/image, `commitLayer` does do a layer presence check before deciding to create it — but in case it is reusing an existing local layer by extracting it into a temporary tarball to be applied, there is still quite a window in which the layer could be concurrently created. Of course, c/image can add one more check to its caller — but if we happened to take locks to read the parent’s state, a lookup for an ID already existing would be ~free.)

Luap99 commented on storage/internal/tempdir/tempdir.go:
```diff
@@ -190,6 +214,23 @@ func NewTempDir(rootDir string) (*TempDir, error) {
+// the temporary location to the final one.
+//
+// The caller MUST ensure .Cleanup() is called after Commit() otherwise the staged content
+// will be deleted and the move will fail.
```
Maybe the wording is confusing, what I tried to say is Commit() must be called before Cleanup() on the TempDir otherwise the staged content is removed and commit fails. Oh how I would love to have this written with rust lifetimes.

Luap99 commented:
```shell
#!/usr/bin/env bpftrace

kfunc:fcntl_setlk
{
  $lockname = str(args->filp->f_path.dentry->d_name.name);

  if ($lockname == "storage.lock" || $lockname == "layers.lock" ||
      $lockname == "images.lock" || $lockname == "containers.lock" ) {

    @blockedTime[tid, $lockname] = nsecs;
  }
}

kretfunc:fcntl_setlk
{
  $lockname = str(args->filp->f_path.dentry->d_name.name);

  if ($lockname == "storage.lock" || $lockname == "layers.lock" ||
      $lockname == "images.lock" || $lockname == "containers.lock" ) {

    // lock duration in msec
    $lock_duration = (nsecs - @blockedTime[tid, $lockname])/1000000;
    if ($lock_duration) {
      printf("blocked %s time: %lld msec\n", $lockname, $lock_duration);
    }
    delete(@blockedTime[tid, $lockname]);
    @lockholdTime[pid, args->fd] = (nsecs, $lockname);
  }
}

tracepoint:syscalls:sys_enter_close
{
  $a = @lockholdTime[pid, args->fd];
  if ($a.0) {
    // lock duration in msec
    $lock_duration = (nsecs - $a.0)/1000000;
    if ($lock_duration) {
      printf("lock %s time: %lld msec\n", $a.1, $lock_duration);
    }
    delete(@lockholdTime[pid, args->fd]);
  }
}
```

FYI this is the script I used to measure lock times during my demo last week, not sure if there is a decent place to store this. I guess it could be helpful in the future to find more lock contention, should I add it under hack/ maybe?

mtrmac commented on storage/drivers/overlay/overlay.go:
```diff
@@ -1013,22 +1056,15 @@ func (d *Driver) create(id, parent string, opts *graphdriver.CreateOpts, readOnl
 
-	st := idtools.Stat{IDs: idPair, Mode: defaultPerms}
-
 	if parent != "" {
```
This part was copied into `getLayerPermissions`, so I think it can be deleted here.

mtrmac commented on storage/layers.go:
```diff
@@ -2481,62 +2626,101 @@ func (r *layerStore) applyDiffWithOptions(to string, layerOptions *LayerOptions,
+		return -1, err
+	}
+
+	if err := tarSplitFile.Sync(); err != nil {
```
(If we do add error checking to `.Close()`, maybe `FdataSync` would suffice — that’s what `AtomicWriteFile` did, at least. That would require exporting the per-platform code in `pkg/ioutils`. Applies throughout.)

mtrmac commented on storage/layers.go:
```diff
@@ -1568,15 +1628,28 @@ func (r *layerStore) create(id string, parentLayer *Layer, names []string, mount
+			}
 		}
 	} else {
 		// applyDiffWithOptions() would have updated r.bycompressedsum
```
Pre-existing, but this should refer to all of `slo` (or whatever it ends up being named) nowadays.

mtrmac commented on storage/store.go:
```diff
@@ -1452,7 +1452,40 @@ func (s *store) canUseShifting(uidmap, gidmap []idtools.IDMap) bool {
+			defer lstore.stopReading()
+		}
+		if l, err := lstore.Get(parent); err == nil && l != nil {
+			ilayer = l
```
(As is, this could `return l` immediately`, but first see elsewhere about keeping the parent store locked.)

mtrmac commented on storage/store.go:
```diff
@@ -1533,19 +1545,107 @@ func (s *store) putLayer(rlstore rwLayerStore, rlstores []roLayerStore, id, pare
+		if !compareIdMappings(freshLayer.UIDMap, parentLayer.UIDMap) || !compareIdMappings(freshLayer.GIDMap, parentLayer.GIDMap) {
+			// Fatal problem. Mappings changed so the parent must be considered different now.
+			// Since we consumed the diff there is no we to recover, return error to caller. The caller would need to retry.
+			// How likely is that and would need to return a special error so c/image could do the retries?
```
(This should currently not happen for c/image-created layers, because those have content-based IDs. It could happen only with a non-c/image writer that does something different. But I don’t know whether we will continue to have deterministic IDs with the sha512 work, I suspect we might not.

Either way, c/image pulls are currently not resilient to concurrent layer removals, so a hard failure here is ~fine. We’d need some kind of “lease a layer” feature that allows the pull to hold a found parent layer while downloading a child, and that does not exist.)

mtrmac commented on storage/store.go:
```diff
@@ -1533,19 +1545,107 @@ func (s *store) putLayer(rlstore rwLayerStore, rlstores []roLayerStore, id, pare
+		if err != nil {
+			return nil, -1, err
+		}
+		if !compareIdMappings(freshLayer.UIDMap, parentLayer.UIDMap) || !compareIdMappings(freshLayer.GIDMap, parentLayer.GIDMap) {
```
This is fine, but it’s a somewhat remote dependency on `populateLayerOptions`; I’d like some fairly explicit comments linking this condition and the code in `populateLayerOptions`… or maybe even wrapping this condition into `populateLayerOptionsInputsUnchanged` function.

```go
struct populateLayerOptionsInput { uidMap, gidMap }
func (… populateLayerOptionsInput) equals(populateLayerOptionsInput) bool {…}
func populateLayerOptionsInputFromLayer(*Layer) populateLayerOptionsInput {…}
// only consume parentLayer via populateLayerOptionsInput(parentLayer) in populateLayerOptionsInput
```
is way too overengineered and not worth the mental overhead, but I think we need _something_.

mtrmac commented on storage/drivers/template.go:
```diff
@@ -1,52 +0,0 @@
-package graphdriver
```
(Unrelated: It turns out there are ~no non-trivial implementations of `CreateFromTemplate` now, so that method could be deleted and the caller simplified a bit not to compute its extra parameters. But that’s way out of scope.)

mtrmac commented on storage/store.go:
```diff
@@ -1533,19 +1545,107 @@ func (s *store) putLayer(rlstore rwLayerStore, rlstores []roLayerStore, id, pare
+			// How likely is that and would need to return a special error so c/image could do the retries?
+			return nil, -1, fmt.Errorf("error during staged layer apply, parent layer %q changed id mappings while the content was extracted, must retry layer creation", parent)
+		}
+		// FIXME: do we need to validate something else that would affect the extracted content?
```
This is fine, but it’s a somewhat remote dependency on `populateLayerOptions`; I’d like some fairly explicit comments linking the ID mapping comparison condition and the code in `populateLayerOptions`… or maybe even wrapping this condition into `populateLayerOptionsInputsUnchanged` function.

```go
struct populateLayerOptionsInput { uidMap, gidMap }
func (… populateLayerOptionsInput) equals(populateLayerOptionsInput) bool {…}
func populateLayerOptionsInputFromLayer(*Layer) populateLayerOptionsInput {…}
// only consume parentLayer via populateLayerOptionsInput(parentLayer) in populateLayerOptionsInput
```
would be way too overengineered and not worth the mental overhead, but I think we need _something_.

mtrmac commented on storage/drivers/overlay/overlay.go:
```diff
@@ -2369,31 +2407,88 @@ func (d *Driver) DifferTarget(id string) (string, error) {
+		return t.Cleanup, nil, -1, err
+	}
+
+	_, forcedSt, st, err := d.getLayerPermissions(parent, options.Mappings.UIDs(), options.Mappings.GIDs())
```
Please propagate the comment about checking for parent existence also to `stageWithUnlockedStore`.

mtrmac commented on storage/layers.go:
```diff
@@ -2398,37 +2473,104 @@ func (r *layerStore) ApplyDiff(to string, diff io.Reader) (size int64, err error
+		cleanup, stageLayer, size, err := sl.staging.StartStagingDiffToApply(parent, drivers.ApplyDiffOpts{
+			Diff:     payload,
+			Mappings: idtools.NewIDMappingsFromMaps(layerOptions.UIDMap, layerOptions.GIDMap),
+		})
```
Fair enough.

mtrmac commented on storage/drivers/overlay/overlay.go:
```diff
@@ -2369,31 +2391,94 @@ func (d *Driver) DifferTarget(id string) (string, error) {
+		// If we have a ENOENT it means the parent was removed which can happen as we are unlocked here.
+		// In this case also wrap ErrLayerUnknown which some callers can handle to retry after recreating the parent.
+		if errors.Is(err, fs.ErrNotExist) {
+			return t.Cleanup, nil, -1, fmt.Errorf("parent layer %q: %w: %w", parent, graphdriver.ErrLayerUnknown, err)
```
(Absolutely non-blocking: `err = fmt.Errorf(…)` and sharing the `return` below would be a tiny bit shorter.)

mtrmac commented on storage/layers.go:
```diff
@@ -288,7 +331,12 @@ type rwLayerStore interface {
+	// checkIdOrNameConfict checks if the id or names are already in use and returns an
+	// error in that case. As Special case if the layer already exists it returns it as
+	// well together with the error.
+	checkIdOrNameConfict(id string, names []string) (*Layer, error)
```
This is now called only with the read lock, so it should be in `roLayerStore`.

mtrmac commented on storage/layers.go:
```diff
@@ -2398,37 +2473,108 @@ func (r *layerStore) ApplyDiff(to string, diff io.Reader) (size int64, err error
+	if err != nil {
+		return err
+	}
+	defer f.Close()
```
This has no error checking, and unlike `applyDiffWithOptions`, it does not call `Sync`.

… so I’m tempted to return to the previous discussion, and prefer error checking when closing the tar-split (in both code paths), just to keep to the more idiomatic structures.

mtrmac commented on storage/layers.go:
```diff
@@ -2481,62 +2629,101 @@ func (r *layerStore) applyDiffWithOptions(to string, layerOptions *LayerOptions,
+	return result.size, err
+}
+
+func applyDiffResultToLayer(r *layerStore, layer *Layer, layerOptions *LayerOptions, result *applyDiffResult) {
```
```suggestion
// Requires startWriting.
func applyDiffResultToLayer(r *layerStore, layer *Layer, layerOptions *LayerOptions, result *applyDiffResult) {
```
and perhaps make it a method on `layerStore`, when it is updating its state fields.

mtrmac commented on storage/store.go:
```diff
@@ -1452,9 +1452,51 @@ func (s *store) canUseShifting(uidmap, gidmap []idtools.IDMap) bool {
 // - rlstore must be locked for writing
 // - rlstores MUST NOT be locked
-func (s *store) putLayer(rlstore rwLayerStore, rlstores []roLayerStore, id, parent string, names []string, mountLabel string, writeable bool, lOptions *LayerOptions, diff io.Reader, slo *stagedLayerOptions) (*Layer, int64, error) {
+func (s *store) putLayer(rlstore rwLayerStore, id string, parentLayer *Layer, names []string, mountLabel string, writeable bool, options *LayerOptions, lc *layerCreationContents) (*Layer, int64, error) {
```
(Non-blocking: Maybe this can be inlined into the callers now.)

mtrmac commented on storage/store.go:
```diff
@@ -1452,9 +1452,51 @@ func (s *store) canUseShifting(uidmap, gidmap []idtools.IDMap) bool {
 // On entry:
 // - rlstore must be locked for writing
 // - rlstores MUST NOT be locked
```
This is not true any more … in fact we do want the store containing the parent, if any, to be locked. See elsewhere about inlining instead.

mtrmac commented on storage/store.go:
So I was wrong about locking here - this call requires holding the store locked for writing, and all callers of `populateLayerOptions` will need a write lock.

mtrmac commented on storage/store.go:
```diff
@@ -1452,9 +1452,51 @@ func (s *store) canUseShifting(uidmap, gidmap []idtools.IDMap) bool {
+// Returns the new copied LayerOptions with mappings set, the parent Layer and
+// an extra unlock function to unlock any potentially read locked rlstores by this function.
+// The unlock function is always set and thus must always be called.
+func populateLayerOptions(s *store, rlstore roLayerStore, rlstores []roLayerStore, parent string, lOptions *LayerOptions) (*LayerOptions, *Layer, func(), error) {
```
At least
```suggestion
func populateLayerOptions(s *store, rlstore roLayerStore, rlstores []roLayerStore, parent string, lOptions *LayerOptions) (*LayerOptions, *Layer, func(), error) {
  // WARNING: Update also the freshLayer checks in store.PutLayer if adding more logic here.
```
please, for https://github.com/containers/container-libs/pull/378#discussion_r2528956792 .

mtrmac reviewed (commented):
A full review (and the read of the resulting code as a whole) now.

Luap99 commented:
@mtrmac I addressed all your comments now. Podman is also happy in https://github.com/containers/podman/pull/27251 (modulo some unrelated flakes)
